### PR TITLE
Review fixes: autograd correctness, optimizer/trainer robustness

### PR DESCRIPTION
**Pull request by workofart**
## Summary

A batch of 18 correctness and robustness fixes across the autograd core, NN modules, optimizer, trainer, and examples:

**Autograd / tensor core**
- Fix `Stack.backward` dropping grads after a non-requires-grad input
- Fix backward crash for `sum()` on 0-d tensors
- Make `Tensor.grad` assignment set instead of accumulate

**NN modules**
- Fix `is_causal` being ignored when an explicit mask is supplied
- Fix Xavier init fan computation for conv kernels
- Fix `Module.zero_grad` being a silent no-op
- Make registered parameters readable as module attributes
- Make `Module.__init__` strict; stop layers swallowing constructor args

**Optimizer / DDP**
- Fix DDP gradient scaling/clipping happening before the AllReduce
- Fix `Adam.load_state_dict` dropping defaultdict and master invariants
- Make TestAdam step/zero_grad tests actually exercise the optimizer

**Trainer / checkpointing**
- Fix final metrics row never being persisted
- Fail fast on `checkpoint_freq` not aligned with `report_every_steps`
- Make `checkpoint_path` an explicit trainer parameter

**Data / examples**
- Validate tokenizer vocab on load; key encoded cache by vocab fingerprint
- Use a sequential sampler for the SFT validation loader
- Evaluate seq2seq example on the test split, not training data
- Add `autograd/tools/__init__.py` so the package ships in builds

**README update**
- Include latest post-training techniques and include inference example with pre-trained checkpoint

## Test plan
- [x] PyTorch-based unit tests pass (correctness validation per repo guidance)

**Dependencies**
- Bump `idna` 3.11 → 3.18 in `uv.lock` and `docs/requirements.txt`, resolving both open Dependabot alerts (medium: CVE-2024-3651 fix bypass in `idna.encode()`)

### CODE REVIEW

File: autograd/text/tokenizer.py
Comment (workofart) [inline autograd/text/tokenizer.py:1]: mostly to prevent mis-reading another encoded data file and silently train on it

File: autograd/tools/trainer.py [hunk 1]
```diff
@@ -571,8 +575,17 @@ def optimizer_step(
...
-        grad_scale = 1.0 / xp.maximum(
-            state.accumulated_loss_total_weight,
+        # Synchronize across DDP ranks BEFORE scaling/clipping so the token
+        # normalization and the clip both operate on the global gradient.
+        # allreduce_grads averages the per-rank gradient sums, so multiply by
+        # world_size to recover the global sum before dividing by the global
+        # token count: world_size * mean(grad_sums) / sum(token_counts)
+        # == sum(grad_sums) / sum(token_counts). Both collectives no-op when
+        # world_size == 1.
+        allreduce_grads(self.optimizer.model_parameters)
```
Comment (workofart) [inline autograd/tools/trainer.py:585]: calling `allreduce_grads` is the responsibility of the trainer, not the `optimizer.step()`. The optimizer doesn't need to know about distributed training.

File: autograd/tensor.py [hunk 2]
```diff
@@ -274,44 +274,26 @@ def grad(self) -> Optional["Tensor"]:
...
 
     @grad.setter
     def grad(self, value: Union["Tensor", ArrayLike, None]) -> None:
         """
-        Set or accumulate gradient for this tensor.
+        Set the gradient for this tensor, replacing any existing gradient.
 
-        If the tensor does not have a gradient yet, we initialize it with the provided value.
-        Otherwise, the provided value is added in place to the existing gradient.
+        Gradient ACCUMULATION during backprop goes through _accumulate_grad;
```
Comment (workofart) [inline autograd/tensor.py:284]: gradient accumulation handled on line 806: https://github.com/workofart/ml-by-hand/pull/93/changes#diff-bc6026b6efb9c0c9beffa739b096273d248b2bd6578d316675aea13d3c03deb6R806

File: examples/sft_gpt_2.py [hunk 3]
```diff
@@ -235,6 +242,7 @@ def generate_eval_samples(
         max_tokens=max_tokens,
         pad_idx=pad_idx,
         sort_buffer_size=CONFIG.eval_batch_size,
+        shuffle=False,
```
Comment (workofart) [inline examples/sft_gpt_2.py:247]: don't shuffle validation dataset
